doc/design/drafts/dynamic_size_types.md

```diff
@@ -0,0 +1,562 @@
+# Dynamic Size Support for Zero Copy Types
+
+## Summary
+
+There are multiple obstacles when it comes to use true zero copy with types
+whose size is not known at compile time. At the moment a developer can pursue
+three approaches.
+
+1. Send via untyped publisher and provide the size at runtime to the publisher.
+   But defining message types with multiple dynamic sized containers would be
+   a real challenge when they should be constructed inside of this sample.
+
+2. One can perform a worst case estimation at compile time and declare
+   a type with container members which fullfil those estimations.
+   This could lead to a lot of wasted memory and if the worst case estimations
+   have to be adjusted all the applications which are using this type may have
+   to be adjusted as well since the capacity is part of the type.
+
+3. Another approach is to use the heap and copy/serialize the data into a sample
+   before sending it with iceoryx. This would imply multiple copies since the
+   publisher has to copy the data from the container into the sample and every
+   subscriber may has to deserialize/copy the data out of the iceoryx sample.
+
+Those obstacles can be overcome when iceoryx containers like `cxx::string`,
+`cxx::vector` or `cxx::{forward}list` support custom allocators which allocate
+memory in a part of the shared memory which is also in the process space of the
+subscriber available.
+
+To accomplish dynamic size types we require:
+
+1. an allocator concept which supports states for allocators
+2. have to adjust the containers like `cxx::string`, `cxx::vector` and 
+   `cxx::{forward}list`
+3. the publisher has to provide access to the shared memory allocator via the
+   sample
+
+### Allocator Type Does Not Effect Container Type
+
+One important restriction for the allocator concept is that the specification
+of the allocator does not change the type like the STL C++ allocator concept
+does. Function developers for instance do not want to restrict their functions
+to shared memory allocated types which would result in some implementation
+overhead. One example could be that someone analyses an image with a function
+like
+`void analyseImage(cxx::vector<int, SharedMemoryAllocator> & a)`. This function
+can now be only used with a vector which is using the `SharedMemoryAllocator`.
+One could define the allocator type as a template but this would increase the
+compile time and can make it nearly impossible to write proprietary code since
+templates require the implementation in the header. Furthermore, the function
+developer has to be aware of the implementation detail that there is something
+like a `SharedMemoryAllocator`.
```

Comment:
Basically the approach of https://en.cppreference.com/w/cpp/memory/polymorphic_allocator but we will have additional complications due to shared memory (across processes).